salvo/src/lib/test_source_tree.py

```diff
@@ -94,50 +251,321 @@ def _check_output_side_effect_fail(*args):
   """Verify that we can identify a failure when attempting to manage commit
      hashes.
   """
-  pass
+  origin = 'https://github.com/someawesomeproject/repo.git'
+  st = _generate_source_tree_from_origin(origin)
+
+  commit_hash = 'invalid_hash_reference'
+
+  mock_check_output.side_effect = _check_output_side_effect_fail
+  hash_string = st.get_previous_commit_hash(commit_hash)
+  assert not hash_string
 
 def test_parent_branch_ahead():
   """Verify that we can determine how many commits beind the local source tree
      lags behind the remote repository.
   """
-  pass
+  origin = 'https://github.com/someawesomeproject/repo.git'
+  st = _generate_source_tree_from_origin(origin)
+
+  gitcmd = 'git status'
+  git_output = """On branch master
+Your branch is ahead of 'origin/master' by 99 commits.
+  (use "git push" to publish your local commits)
+
+nothing to commit, working tree clean
+"""
+  with mock.patch('src.lib.cmd_exec.run_command',
+                  mock.MagicMock(return_value=git_output)) as magic_mock:
+    commit_count = st.get_revs_behind_parent_branch()
+    cmd_params = cmd_exec.CommandParameters(cwd=mock.ANY)
+    magic_mock.assert_called_once_with(gitcmd, cmd_params)
+
+    assert isinstance(commit_count, int)
+    assert commit_count == 99
 
 def test_parent_branch_up_to_date():
   """Verify that we can determine how many commits beind the local source tree
      lags behind the remote repository.
   """
-  pass
+  origin = 'https://github.com/someawesomeproject/repo.git'
+  st = _generate_source_tree_from_origin(origin)
+
+  gitcmd = 'git status'
+  git_output = """On branch master
+Your branch is up to date with 'origin/master'.
+
+Changes not staged for commit:
+  (use "git add <file>..." to update what will be committed)
+  (use "git checkout -- <file>..." to discard changes in working directory)
+"""
+  with mock.patch('src.lib.cmd_exec.run_command',
+                  mock.MagicMock(return_value=git_output)) as magic_mock:
+    commit_count = st.get_revs_behind_parent_branch()
+    cmd_params = cmd_exec.CommandParameters(cwd=mock.ANY)
+    magic_mock.assert_called_once_with(gitcmd, cmd_params)
+
+    assert isinstance(commit_count, int)
+    assert commit_count == 0
 
 def test_branch_up_to_date():
   """Verify that we can determine a source tree is up to date."""
-  pass
+
+  origin = 'https://github.com/someawesomeproject/repo.git'
+  st = _generate_source_tree_from_origin(origin)
+
+  with mock.patch(
+      'src.lib.source_tree.SourceTree.get_revs_behind_parent_branch',
+      mock.MagicMock(return_value=0)) as magic_mock:
+    up_to_date = st.is_up_to_date()
+    magic_mock.assert_called_once()
+    assert up_to_date
+
 
 def test_list_tags():
   """Verify that we can list tags from a repository."""
-  pass
+
+  origin = 'https://github.com/someawesomeproject/repo.git'
+  st = _generate_source_tree_from_origin(origin)
+
+  gitcmd = "git tag --list --sort v:refname"
+  git_output = """v1.0.0
+v1.1.0
+v1.2.0
```

Comment:
Test data are also code in the sense that they need to be understood and maintained with the code changes. Can we minimize this test data (the list of tags) to the bare minimum necessary to test the behavior?

This will decrease the amount of lines we have to maintain here and make it easier to understand the purpose of the test. It isn't clear at the moment if all of these are needed and / or some of them test some special behaviors.

Applies to the remaining two tag lists below as well.

Reply:
Updated